### PR TITLE
Refactor Test Framework

### PR DESCRIPTION
**Pull request by karenc-bq**
### Summary

Refactor test framework so it is more readable and maintainable

### Description

Separate tests into two classes, move shared methods out and change method signatures.

### Additional Reviewers
@sergiyv-bitquill 
@ColinKYuen 
@crystall-bitquill 
@karenc-bq

### CODE REVIEW

**Review by ColinKYuen — APPROVED** [review]
lgtm

**Review by ColinKYuen — APPROVED** [review]

**Review by ColinKYuen — APPROVED** [review]

**Review by ColinKYuen — APPROVED** [review]

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  private static int mySQLProxyPort;
+  private static GenericContainer<?> integrationTestContainer;
+  private static String dbHostCluster = "";
+  private static String dbHostClusterRo = "";
+
+  private Network network;
+
+  @BeforeEach
+  void setUp() {
+    final String dbConnHostBase =
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:73]: Let's add assert for DB_CONN_STR_SUFFIX and TEST_DB_CLUSTER_IDENTIFIER here. It helps to troubleshoot failures.

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        .withEnv("TEST_USERNAME", TEST_USERNAME)
+        .withEnv("TEST_PASSWORD", TEST_PASSWORD)
+        .withEnv("DB_CLUSTER_CONN", dbHostCluster)
+        .withEnv("DB_RO_CLUSTER_CONN", dbHostClusterRo)
+        .withEnv("TOXIPROXY_CLUSTER_NETWORK_ALIAS", "toxiproxy-instance-cluster")
+        .withEnv("TOXIPROXY_RO_CLUSTER_NETWORK_ALIAS", "toxiproxy-ro-instance-cluster")
+        .withEnv("PROXIED_CLUSTER_TEMPLATE", "?" + DB_CONN_STR_SUFFIX + PROXIED_DOMAIN_NAME_SUFFIX);
+
+    // Add mysql instances & proxies to container env
+    for (int i = 0; i < mySqlInstances.size(); i++) {
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:166]: Add mySqlInstances as method parameter?

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+  @Test
+  public void testRunTestInContainer()
+      throws UnsupportedOperationException, IOException, InterruptedException, SQLException {
+    initializeToxiProxy(network, toxiproxyContainerList);
+    integrationTestContainer = initializeTestContainer(network);
+    runTest(integrationTestContainer, "test-integration-container-aurora");
+  }
+
+  private void initializeToxiProxy(Network network, List<ToxiproxyContainer> toxiproxyContainerList) {
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:98]: Add mySqlInstances as parameter?

File: src/test/java/testsuite/integration/host/ContainerBaseTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      throws IOException, InterruptedException {
+    System.out.println("==== Container console feed ==== >>>>");
+    Consumer<OutputFrame> consumer = new ConsoleConsumer();
+    Integer exitCode =
+        ExecInContainerUtility.execInContainer(container, consumer, "./gradlew", task);
+    System.out.println("==== Container console feed ==== <<<<");
+    assertEquals(0, exitCode, "Some tests failed.");
+  }
+
+  static GenericContainer<?> createTestContainerTemplate(String dockerImageName) {
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/ContainerBaseTest.java:62]: What's the reason to make this method static? It seems like just instance protected method.

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class AuroraIntegrationContainerTest extends ContainerBaseTest {
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:49]: I'd like to see class header with a list of environment variables and sample values for them.

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    final String dbConnHostBase =
+        DB_CONN_STR_SUFFIX.startsWith(".")
+            ? DB_CONN_STR_SUFFIX.substring(1)
+            : DB_CONN_STR_SUFFIX;
+    dbHostCluster = TEST_DB_CLUSTER_IDENTIFIER + ".cluster-" + dbConnHostBase;
+    dbHostClusterRo = TEST_DB_CLUSTER_IDENTIFIER + ".cluster-ro-" + dbConnHostBase;
+    network = Network.newNetwork();
+  }
+
+  @AfterEach
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:82]: @BeforeAll ?
Test environment (containers) can be created once and can be reused by different tests. For this particular case it doesn't make a big difference but as a pattern for further environment it makes sense.

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  private static final List<String> mySqlInstances = new ArrayList<>();
+
+  private static int mySQLProxyPort;
+  private static GenericContainer<?> integrationTestContainer;
+  private static String dbHostCluster = "";
+  private static String dbHostClusterRo = "";
+
+  private Network network;
+
+  @BeforeEach
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:71]: @BeforeAll ?

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+      System.getenv("TEST_DB_CLUSTER_IDENTIFIER");
+  private static final String PROXIED_DOMAIN_NAME_SUFFIX = ".proxied";
+  private static final String RETRIEVE_TOPOLOGY_SQL =
+      "SELECT SERVER_ID FROM information_schema.replica_host_status ";
+  private static final DockerImageName TOXIPROXY_IMAGE =
+      DockerImageName.parse("shopify/toxiproxy:2.1.0");
+  private static final List<ToxiproxyContainer> toxiproxyContainerList = new ArrayList<>();
+  private static final List<String> mySqlInstances = new ArrayList<>();
+
+  private static int mySQLProxyPort;
```
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:64]: Why static?

File: src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java (hunk 3 above)
Comment (sergiyv-bitquill) [inline src/test/java/testsuite/integration/host/AuroraIntegrationContainerTest.java:98]: protected ?
It helps to create other aurora-based test environments by inheriting from this class.

File: src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
+package testsuite.integration.container;
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java:1]: Licence

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/NodeMonitoringConnectionPlugin.java [hunk 10]
```diff
@@ -293,7 +295,15 @@ private void checkIfChanged(JdbcConnection newConnection) {
    */
   private void generateNodeKeys(Connection connection) {
     this.nodeKeys.clear();
+
     final HostInfo hostInfo = this.currentConnectionProvider.getCurrentHostInfo();
+    this.nodeKeys.add(
```
Comment (ColinKYuen) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/NodeMonitoringConnectionPlugin.java:300]: Why is this moved up? 
Originally it was at the bottom to have the `IP & port` combo first in the set/list.

Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/NodeMonitoringConnectionPlugin.java:300]: No particular reason. Just move it closer to hostinfo. Since nodeKeys is a Set<> there's no difference in what order you add values to it.

File: src/test/java/testsuite/integration/utility/ContainerHelper.java [hunk 11]
```diff
@@ -0,0 +1,281 @@
+package testsuite.integration.utility;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.ExecCreateCmdResponse;
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.exception.DockerException;
+import eu.rekawek.toxiproxy.Proxy;
+import eu.rekawek.toxiproxy.model.ToxicDirection;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.containers.*;
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/utility/ContainerHelper.java:10]: Missing license. Star imports

File: src/test/java/testsuite/integration/host/CommunityContainerTest.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+  private static final String TEST_CONTAINER_NAME = "test-container";
+  private static final String MYSQL_CONTAINER_NAME = "mysql-container";
+
+  private static MySQLContainer<?> communityMysqlContainer;
+  private static GenericContainer<?> communityTestContainer;
+  private static Network network;
+  private static final ContainerHelper containerHelper = new ContainerHelper();
+
+
+  @BeforeAll
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/host/CommunityContainerTest.java:52]:
```suggestion

  @BeforeAll
```

File: src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+      cell = row.createCell(2);
+      cell.setCellValue("FailureDetectionCount");
+      cell = row.createCell(3);
+      cell.setCellValue("NetworkOutageDelayMillis");
+      cell = row.createCell(4);
+      cell.setCellValue("MinFailureDetectionTime");
+      cell = row.createCell(5);
+      cell.setCellValue("MaxFailureDetectionTime");
+      cell = row.createCell(6);
+      cell.setCellValue("AvgFailureDetectionTime");
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java:59]:
```suggestion
      final String[] headers = new String[]{
          "FailureDetectionGraceTime",
          "FailureDetectionInterval",
          "FailureDetectionCount",
          "NetworkOutageDelayMillis",
          "MinFailureDetectionTime",
          "MaxFailureDetectionTime",
          "AvgFailureDetectionTime"
      };

      // Header
      Row row = sheet.createRow(0);
      Cell cell;
      for (int i = 0; i < headers.length; i++) {
        cell = row.createCell(i);
        cell.setCellValue(headers[i]);
      }
```

File: src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        cell = row.createCell(2);
+        cell.setCellValue(perfStat.paramDetectionCount);
+        cell = row.createCell(3);
+        cell.setCellValue(perfStat.paramNetworkOutageDelayMillis);
+        cell = row.createCell(4);
+        cell.setCellValue(perfStat.minFailureDetectionTimeMillis);
+        cell = row.createCell(5);
+        cell.setCellValue(perfStat.maxFailureDetectionTimeMillis);
+        cell = row.createCell(6);
+        cell.setCellValue(perfStat.avgFailureDetectionTimeMillis);
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java:79]: same as above?

File: src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+  protected final ContainerHelper containerHelper = new ContainerHelper();
+
+  @BeforeAll
+  public static void setUp() throws IOException, SQLException {
+    toxiproxyClientInstance_1 = new ToxiproxyClient(TOXIPROXY_INSTANCE_1_NETWORK_ALIAS, TOXIPROXY_CONTROL_PORT);
+    toxiproxyClientInstance_2 = new ToxiproxyClient(TOXIPROXY_INSTANCE_2_NETWORK_ALIAS, TOXIPROXY_CONTROL_PORT);
+    toxiproxyClientInstance_3 = new ToxiproxyClient(TOXIPROXY_INSTANCE_3_NETWORK_ALIAS, TOXIPROXY_CONTROL_PORT);
+    toxiproxyClientInstance_4 = new ToxiproxyClient(TOXIPROXY_INSTANCE_4_NETWORK_ALIAS, TOXIPROXY_CONTROL_PORT);
+    toxiproxyClientInstance_5 = new ToxiproxyClient(TOXIPROXY_INSTANCE_5_NETWORK_ALIAS, TOXIPROXY_CONTROL_PORT);
```
Comment (karenc-bq) [inline src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java:72]: Do we actually need all these? Can we keep them in an array?

Comment (sergiyv-bitquill) [reply src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java:72]: I was thinking about it and I'm also not a big fun of the way we pass environment variables to a container and how we load them up in a container. This area requires some re-view and re-design. For the moment I'd keep it as is.

File: src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java (hunk 13 above)
Comment (sergiyv-bitquill) [reply src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java:59]: I'm not sure if this changes make the code more readable.

File: src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java (hunk 15 above)
Comment (karenc-bq) [reply src/test/java/testsuite/integration/container/AuroraMysqlIntegrationBaseTest.java:103]: Sounds good. Created a ticket so we don't forget: https://bitquill.atlassian.net/browse/RDS-625

File: src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java (hunk 13 above)
Comment (karenc-bq) [reply src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java:90]: We can hide it in a method. The original method drives me crazy.

Comment (karenc-bq) [reply src/test/java/testsuite/integration/container/AuroraMysqlPerformanceIntegrationTest.java:90]: `buildCSVHeaders(String[] headers)`?
